metric-publishers/emf-metric-publisher/src/main/java/software/amazon/awssdk/metrics/publishers/emf/MetricEmfConverter.java

```diff
@@ -0,0 +1,368 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.metrics.publishers.emf;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.metrics.MetricCategory;
+import software.amazon.awssdk.metrics.MetricCollection;
+import software.amazon.awssdk.metrics.MetricRecord;
+import software.amazon.awssdk.protocols.jsoncore.JsonWriter;
+import software.amazon.awssdk.utils.Logger;
+
+/**
+ * # MetricEmfConverter
+
+ * Converts metrics into Amazon CloudWatch Embedded Metric Format (EMF).
+ * This internal class handles the transformation of various metric types
+ * into EMF-compatible format.
+
+ * ## Configuration
+ * The converter is initialized with an `EmfMetricConfiguration` that defines:
+ * - Metric categories to process
+ * - Other EMF-specific settings
+
+ * ## Implementation Notes
+ * - Boolean values are converted to numeric values (1.0 for true, 0.0 for false)
+ * - Null values are converted to 0.0
+ * - Very small numeric values (below 0.0001) are normalized to 0.0
+ * - Duration values are converted to milliseconds
+ *
+ * @since 2.0
+ * @see EmfMetricConfiguration
+ */
+@SdkInternalApi
+public class MetricEmfConverter {
```

Comment:
Missing unit tests for this class

Reply:
So should I test the converting method in this class and the publish method in the main class?

Reply:
Yes, ideally, each class should have its own test focusing on its own functionality.